### PR TITLE
Implement CounterFactory, which is used by ZuulException

### PR DESCRIPTION
**Pull request by Aloren**
`ZuulException` has following code:
```
    public ZuulException(String sMessage, int nStatusCode, String errorCause) {
        super(sMessage);
        this.nStatusCode = nStatusCode;
        this.errorCause = errorCause;
        incrementCounter("ZUUL::EXCEPTION:" + errorCause + ":" + nStatusCode);
    }
```
Creating exception increments metrics counter. Before this pull request this feature was disabled:
 see `ZuulFilterInitializer`.

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java [hunk 1]
```diff
@@ -20,9 +20,8 @@
 import java.util.Map;
 
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.actuate.metrics.CounterService;
+import org.springframework.boot.autoconfigure.condition.*;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: We don't use star imports.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package org.springframework.cloud.netflix.zuul;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java]: Please add license header

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactory.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package org.springframework.cloud.netflix.zuul.metrics;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactory.java]: Please add license header

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyCounterFactory.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package org.springframework.cloud.netflix.zuul.metrics;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyCounterFactory.java]: Please add license header

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyTracerFactory.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package org.springframework.cloud.netflix.zuul.metrics;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyTracerFactory.java]: Please add license header

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactoryTests.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package org.springframework.cloud.netflix.zuul.metrics;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactoryTests.java]: Please add license header

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/metrics/ZuulEmptyMetricsApplicationTests.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
+package org.springframework.cloud.netflix.zuul.metrics;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/metrics/ZuulEmptyMetricsApplicationTests.java]: Please add license header

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/metrics/ZuulMetricsApplicationTests.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
+package org.springframework.cloud.netflix.zuul.metrics;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/metrics/ZuulMetricsApplicationTests.java]: Please add license header

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+
+import com.netflix.zuul.monitoring.CounterFactory;
+import com.netflix.zuul.monitoring.TracerFactory;
+
+import lombok.extern.apachecommons.CommonsLog;
+
+@CommonsLog
+public class ZuulMetricsInitializer implements ServletContextListener {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java]: I'd rather have once class that does this. Any reason to not include this in `ZuulFilterInitializer`?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactory.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+import org.springframework.boot.actuate.metrics.CounterService;
+
+import com.netflix.zuul.monitoring.CounterFactory;
+
+public class DefaultCounterFactory extends CounterFactory {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactory.java]: Short javadoc at least.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyCounterFactory.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+
+import com.netflix.zuul.monitoring.CounterFactory;
+
+public class EmptyCounterFactory extends CounterFactory {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyCounterFactory.java]: Short javadoc at least.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyTracerFactory.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.netflix.zuul.monitoring.Tracer;
+import com.netflix.zuul.monitoring.TracerFactory;
+
+public class EmptyTracerFactory extends TracerFactory {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/EmptyTracerFactory.java]: Short javadoc at least.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java (hunk 9 above)
Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java]: I don't understand how it relates to filters initialization, so I decided to put it into separate class. I can move it back, but `ZuulFilterInitializer` will know about `CounterFacatory` and `TracerFactory`, I don't think this is actually a good idea. What do you think?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java (hunk 2 above)
Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java]: done

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactory.java (hunk 3 above)
Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/metrics/DefaultCounterFactory.java]: done

**Comment by Aloren** [comment]
Fixed all issues, except for `ZuulMetricsInitializer `.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java (hunk 9 above)
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulMetricsInitializer.java]: Move it there, please. I'd rather have one place to initialize zuul.

**Comment by Aloren** [comment]
Shouldn't `ZuulFilterInitializer` set TracerFactory's and CounterFactory's singletones to null in `contextDestroyed` method?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java [hunk 13]
```diff
@@ -179,8 +186,38 @@ public ZuulRouteApplicationContextInitializer zuulRoutesApplicationContextInitia
...
+		public TracerFactory tracerFactory() {
+			return new EmptyTracerFactory();
+		}
+
+	}
+
+	@Configuration
+	protected static class ZuulMetricsConfiguration {
+
+		@ConditionalOnClass(CounterService.class)
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: I'd move this to the class and add a conditional on missing class on the empty one as well.

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/ZuulFilterInitializerTests.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
+package org.springframework.cloud.netflix.zuul;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/ZuulFilterInitializerTests.java]: License header

**Comment by spencergibb** [comment]
Probably should be reset in destroy.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java (hunk 13 above)
Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: I'm not able to add `@ConditionalOnMissingClass("org.springframework.boot.actuate.metrics.CounterService")` to `ZuulEmptyMetricsConfiguration`, because it is present on classpath in tests. So none of 2 configurations get setup. To fix tests I would need to add `metrics-core` to classpath to enable `DropwizardAutoConfiguration`, so that `ZuulMetricsConfiguration` got setup.

**Comment by Aloren** [comment]
Looks like I've resolved all comments. Thanks for review!

**Comment by Aloren** [comment]
There were issues with tests because of nasty singletones in `ZuulFilterInitializer`. Hope this is not an issue :)

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java [hunk 15]
```diff
@@ -179,8 +188,40 @@ public ZuulRouteApplicationContextInitializer zuulRoutesApplicationContextInitia
 		private Map<String, ZuulFilter> filters;
 
 		@Bean
-		public ZuulFilterInitializer zuulFilterInitializer() {
-			return new ZuulFilterInitializer(this.filters);
+		public ZuulFilterInitializer zuulFilterInitializer(
+				CounterFactory counterFactory, TracerFactory tracerFactory) {
+			FilterLoader filterLoader = FilterLoader.getInstance();
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: Why did you move this out. Seems unreleated.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java [hunk 16]
```diff
@@ -179,8 +188,40 @@ public ZuulRouteApplicationContextInitializer zuulRoutesApplicationContextInitia
...
+			FilterRegistry filterRegistry = FilterRegistry.instance();
+			return new ZuulFilterInitializer(this.filters, counterFactory, tracerFactory, filterLoader, filterRegistry);
+		}
+
+	}
+
+	@Configuration
+	protected static class ZuulEmptyMetricsConfiguration {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: This still needs a conditional on missing class.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java (hunk 15 above)
Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: Because of unit tests. Initializing these singletons breaks spring aware tests.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: OK

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java (hunk 16 above)
Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: Adding @ConditionalOnMissingClass will prevent creating of this configuration in tests because CounterService actually exists in classpath (spring-boot-starter-actuator is optional dependency).

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: Even with `@ClassPathExclusions`?

Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: I would need to add that to all spring aware tests.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: I'm confused. You should only need it on the test that is testing that the default empty configuration works.

Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: There are a lot of tests that use `ZuulConfiguration`, it has inside: `ZuulFilterConfiguration`, which initializes: `ZuulFilterInitializer`. Now `ZuulFilterInitializer` depends on CounterFactory and TracerFactory beans. Both of them are setup either in `ZuulMetricsConfiguration` or in `ZuulEmptyMetricsConfiguration`. 
`ZuulMetricsConfiguration` gets setup only when class CounterService is on classpath (it creates DefaultCounterFactory). 
`ZuulEmptyMetricsConfiguration` gets setup now in any case (it creates EmptyCounterFactory and EmptyTracerFactory).
If I add @ConditionalOnMissingClass(CounterService) to `ZuulEmptyMetricsConfiguration` it won't setup, because CounterService is on classpath. 
So I end up with only `ZuulMetricsConfiguration`, which creates  DefaultCounterFactory and does not create TracerFactory. 

I can add @ConditionalOnMissingClass(CounterService) to `ZuulEmptyMetricsConfiguration`, but I will need to add setup of EmptyTracerFactory in it. Do you agree?

Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: Moreover I will need to add `metrics-core` for `DropwizardAutoConfiguration` to be setup, because of this:
```
		@ConditionalOnBean(CounterService.class)
		@Bean
		public CounterFactory counterFactory(CounterService counterService) {
			return new DefaultCounterFactory(counterService);
		}
```
Maybe there exists some other way of conditional setup of these beans?

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: This should always work. Should the `EmptyTracerFactory` be in it's own configuration then>

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java [hunk 17]
```diff
@@ -179,8 +185,35 @@ public ZuulRouteApplicationContextInitializer zuulRoutesApplicationContextInitia
...
+				CounterFactory counterFactory, TracerFactory tracerFactory) {
+			FilterLoader filterLoader = FilterLoader.getInstance();
+			FilterRegistry filterRegistry = FilterRegistry.instance();
+			return new ZuulFilterInitializer(this.filters, counterFactory, tracerFactory, filterLoader, filterRegistry);
+		}
+
+	}
+
+	@Configuration
+	protected static class ZuulMetricsConfiguration {
```
Comment (Aloren) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulConfiguration.java]: @spencergibb How about only one `ZuulMetricsConfiguration`? I've joined empty and default configurations. Is that ok for you?

**Comment by spencergibb** [comment]
Squashed and merged with a slight change to the conditional on class via 5440aca212a4df95fa37210d38cac200cbfccf4e
